File: implementation1/entities/player.py
from dataclasses import dataclass
from typing import Optional
from common_types import BombInfo, WorldInfo, Direction, EntityType, UpdateResultInfo
from helpers.event import UpdateResult
from helpers.grid_adapter import GridAdapter


@dataclass
class Player():
    def __init__(self, x: float, y: float, world: WorldInfo, grid: GridAdapter, id: int):
        self._x: float = x
        self._y: float = y
        self._world: WorldInfo = world
        self._grid: GridAdapter = grid
        self._id: int = id
        self._speed: float = 2.0
        self._expired: bool = False
        self._alive: bool = True
        self._active_bombs: set[BombInfo] = set()
        self.direction_facing: Direction = Direction.SOUTH
        self._snap_tolerance: int = 2  
        self._hitbox_width = 16
        self._hitbox_height = 16
        self._sprite_width = 16
        self._sprite_height = 24
        self._hitbox_offset_y = self._sprite_height - self._hitbox_height

    @property
    def entity_type(self) -> EntityType:
        return EntityType.PLAYER

    @property
    def is_expired(self) -> bool:
        return self._expired

    @property
    def x(self) -> float:
        return self._x

    @property
    def y(self) -> float:
        return self._y
    
    @property
    def id(self) -> int:
        return self._id

    @property
    def width(self) -> int:
        return self._sprite_width

    @property
    def height(self) -> int:
        return self._sprite_height

    @property
    def speed(self) -> float:
        return self._speed

    @property
    def hitbox_x(self) -> float:
        return self._x

    @property
    def hitbox_y(self) -> float:
        return self._y + self._hitbox_offset_y  # bottom 16x16, like the video (head is non hitbox)

    @property
    def row(self) -> int:
        cx = self.hitbox_x + self._hitbox_width / 2
        cy = self.hitbox_y + self._hitbox_height / 2
        res = self._grid.pixel_to_cell(int(cx), int(cy))
        if res is not None:
            return res[0]
        else:
            return -1

    @property
    def col(self) -> int:
        cx = self.hitbox_x + self._hitbox_width / 2
        cy = self.hitbox_y + self._hitbox_height / 2
        res = self._grid.pixel_to_cell(int(cx), int(cy))
        if res is not None:
            return res[1]
        else:
            return -1

    def set_speed(self, ds: int) -> None:
        self._speed = max(0.0, self._speed + ds)

    def _snap_axis(self, new_x: float, new_y: float, direction: Direction) -> tuple[float, float]:

        #This is for sliding around sa grid, masyado rigid kasi ung pixel perfect
        
        if direction in (Direction.NORTH, Direction.SOUTH):
            cur_row, cur_col = self.row, self.col
            cell_x, cell_y, w, h = self._grid.cell_rect(cur_row, cur_col)
            target_cx = cell_x + w / 2
            new_hb_cx = new_x + self._hitbox_width / 2
            if abs(new_hb_cx - target_cx) <= self._snap_tolerance:
                new_x = target_cx - self._hitbox_width / 2
        
        elif direction in (Direction.WEST, Direction.EAST):
            cur_row, cur_col = self.row, self.col
            cell_x, cell_y, w, h = self._grid.cell_rect(cur_row, cur_col)
            target_cy = cell_y + h / 2
            new_hb_cy = (new_y + self._hitbox_offset_y) + self._hitbox_height / 2 # bottom 16x16
            if abs(new_hb_cy - target_cy) <= self._snap_tolerance:
                new_y = (target_cy - self._hitbox_height / 2) - self._hitbox_offset_y

        return new_x, new_y

    def _can_move_to(self, new_x: float, new_y: float) -> bool:
        hb_x1 = new_x
        hb_y1 = new_y + self._hitbox_offset_y
        hb_x2 = hb_x1 + self._hitbox_width - 1
        hb_y2 = hb_y1 + self._hitbox_height - 1

        corners = [(hb_x1, hb_y1),(hb_x2, hb_y1),(hb_x1, hb_y2),(hb_x2, hb_y2),]

        for px, py in corners:
            res = self._grid.pixel_to_cell(int(px), int(py))
            if res:
                row, col = res
                if self._world.is_cell_blocking(row, col, self.id):
                    return False
            else:
                return False

        return True

    def move(self, direction: Direction) -> int:
        if not self._alive:
            return 0

        dx = dy = 0.0
        if direction == Direction.NORTH:
            dy = -self._speed
        elif direction == Direction.SOUTH:
            dy = self._speed
        elif direction == Direction.WEST:
            dx = -self._speed
        elif direction == Direction.EAST:
            dx = self._speed

        new_x = self._x + dx
        new_y = self._y + dy

        new_x, new_y = self._snap_axis(new_x, new_y, direction)

        if self._can_move_to(new_x, new_y):
            self._x = new_x
            self._y = new_y
            return 1

        return 0

    def handle_input(self, inputs: dict[str, bool]) -> None:
        direction: Optional[Direction] = None

        if inputs.get("UP"):
            direction = Direction.NORTH
        elif inputs.get("DOWN"):
            direction = Direction.SOUTH
        elif inputs.get("LEFT"):
            direction = Direction.WEST
        elif inputs.get("RIGHT"):
            direction = Direction.EAST

        if direction is not None:
            self.direction_facing = direction # kung saan naka harap ung player
            self.move(direction)


    def decide_move(self) -> Direction:
        return Direction.NORTH

    def remove(self, bomb: BombInfo) -> None:
        try:
            self._active_bombs.remove(bomb)
        except ValueError:
            pass

    def add(self, bomb: BombInfo) -> None:
        self._active_bombs.add(bomb)

    def on_explosion_hit(self) -> None:
        if not self._alive:
            return
        self._alive = False
        self._expired = True

    def update(self, dt: int) -> UpdateResultInfo:
        return UpdateResult()


class PlayerFactory:

    @classmethod
    def make(cls, x: float, y: float, world: WorldInfo, grid: GridAdapter, id: int) -> Player:
        return Player(x, y, world, grid, id)

File: implementation2/src/entity/player.ts
import { HashSet } from "effect"
import {
  Model,
  Explosion,
  Bomb,
  Block,
  Player,
  Powerup,
  Entity,
  World,
  UpdateResult,
  Effect,
  DeathSound,
  DeathAnimation,
  AnimationCmd,
  PixelMode,
  GridCoords,
} from "../model";
import { tickEffect } from "./powerup";
import { Array , Option} from "effect";
import { pixelToCell } from "../helpers/grid_adapter";
import { isCellBlocking } from "../helpers/world";

const TILE_SIZE = 16;
const HITBOX_WIDTH = 16;
const HITBOX_HEIGHT = 16;
const HITBOX_OFFSET_Y = 8;
const SNAP_TOLERANCE = 6;


const canMoveTo = (world: World, id: number, x: number, y: number): boolean => {
  const hbX1 = x;
  const hbY1 = y + HITBOX_OFFSET_Y;
  const hbX2 = hbX1 + HITBOX_WIDTH - 1;
  const hbY2 = hbY1 + HITBOX_HEIGHT - 1;

  const corners: [number, number][] = [
    [hbX1, hbY1],
    [hbX2, hbY1],
    [hbX1, hbY2],
    [hbX2, hbY2],
  ];

  for (const [px, py] of corners) {
    const r = Math.floor(py / TILE_SIZE);
    const c = Math.floor(px / TILE_SIZE);
    if (isCellBlocking(world, r, c, id)) {
      return false;
    }
  }
  return true;
};


const snapAxis = (
  x: number,
  y: number,
  vx: number,
  vy: number,
  currentRow: number,
  currentCol: number
): [number, number] => {
  let newX = x;
  let newY = y;

  const cellX = currentCol * TILE_SIZE;
  const cellY = currentRow * TILE_SIZE;

  if (vy !== 0) {
    const targetCX = cellX + TILE_SIZE / 2;
    const currentCX = x + HITBOX_WIDTH / 2;
    
    if (Math.abs(currentCX - targetCX) <= SNAP_TOLERANCE) {
      newX = targetCX - HITBOX_WIDTH / 2;
    }
  }

  else if (vx !== 0) {
    const targetCY = cellY + TILE_SIZE / 2;
    const currentCY = (y + HITBOX_OFFSET_Y) + HITBOX_HEIGHT / 2;
    
    if (Math.abs(currentCY - targetCY) <= SNAP_TOLERANCE) {
       newY = targetCY - HITBOX_HEIGHT / 2 - HITBOX_OFFSET_Y;
    }
  }

  return [newX, newY];
};

const calculateMovement = (player: Player, world: World, dt: number): [number, number] => {
    if (player.vx === 0 && player.vy === 0) {
        return [player.x, player.y];
    }

    const curRow = getPlayerRow(player);
    const curCol = getPlayerCol(player);

    let targetX = player.x + player.vx * dt;
    let targetY = player.y + player.vy * dt;

    [targetX, targetY] = snapAxis(targetX, targetY, player.vx, player.vy, curRow, curCol);

    if (canMoveTo(world, player.player_id, targetX, targetY)) {
        return [targetX, targetY];
    }

    const cellX = curCol * TILE_SIZE;
    const cellY = curRow * TILE_SIZE;
    
    let clampedX = player.x;
    let clampedY = player.y;
    let attemptedClamp = false;

    // East
    if (player.vx > 0) {
        const limitX = (cellX + TILE_SIZE) - HITBOX_WIDTH;
        if (limitX > player.x) {
            clampedX = limitX;
            attemptedClamp = true;
        }
    }
    // West
    else if (player.vx < 0) {
        const limitX = cellX;
        if (limitX < player.x) {
            clampedX = limitX;
            attemptedClamp = true;
        }
    }
    // South
    else if (player.vy > 0) {
        const limitY = (cellY + TILE_SIZE) - HITBOX_HEIGHT - HITBOX_OFFSET_Y;
        if (limitY > player.y) {
            clampedY = limitY;
            attemptedClamp = true;
        }
    }
    // North
    else if (player.vy < 0) {
        const limitY = cellY - HITBOX_OFFSET_Y;
        if (limitY < player.y) {
            clampedY = limitY;
            attemptedClamp = true;
        }
    }

    if (attemptedClamp && canMoveTo(world, player.player_id, clampedX, clampedY)) {
        return [clampedX, clampedY];
    }

    return [player.x, player.y];
}


export const updatePlayer = (
  ent: Player,
  world: World,
  dt: number
): [Player, UpdateResult] => {
  let result = UpdateResult.make({ events: [], sounds: [], animations: [] });
  
  const tickedEffects = ent.effects.map((e) => tickEffect(dt, e));
  const remainingEffects = tickedEffects.filter((e) =>
    Option.getOrElse(e.timeRemaining, () => 1) > 0
  );


  if (!ent.isAlive) {
    let animation = AnimationCmd.make({
      type: DeathAnimation.make(),
      mode: PixelMode.make(),
      a: ent.x,
      b: ent.y,
      durationFrames: (1/dt)*2,
      id: Option.some(ent.id),
      powerupType: Option.none(),
    });
    result = {
      ...result,
      sounds: Array.append(result.sounds, DeathSound.make()),
      animations: Array.append(result.animations, animation),
    };
    return [
      Player.make({ ...ent, effects: remainingEffects, isExpired: true }),
      result,
    ];
  }

  // Handle Movement
  const [nextX, nextY] = calculateMovement(ent, world, 1);
  const centerX = nextX + HITBOX_WIDTH / 2;
  const centerY = nextY + HITBOX_OFFSET_Y + HITBOX_HEIGHT / 2;

  const newRow = Math.floor(centerY / TILE_SIZE);
  const newCol = Math.floor(centerX / TILE_SIZE);
  return [
      Player.make({ 
          ...ent, 
          x: nextX, 
          y: nextY, 
          row: newRow,
          col: newCol,
          effects: remainingEffects 
      }), 
      result
  ];
};

export const onExplosionHitPlayer = (player: Player): Player => {
  if (!player.isAlive) {
    return Player.make({...player})
  }
  return Player.make({...player, isAlive: false})
}

export const playerSpeed = (player: Player): number => Array.reduce(player.effects, player.speed, (acc, effect) => acc + effect.speedDelta)

export const playerRange = (player: Player): number => Array.reduce(player.effects, 1, (acc, effect) => acc + effect.rangeDelta)

export const playerMaxBombs = (player: Player): number => Array.reduce(player.effects, 1, (acc, effect) => acc + effect.bombsDelta)

export const addBomb = (player: Player, bomb: Bomb): Player => Player.make({...player, activeBombs: Array.append(player.activeBombs, bomb.id)})


export const removeBomb = (player: Player, bomb: Bomb): Player => Player.make({...player, activeBombs: Array.filter(player.activeBombs, (id) => id !== bomb.id)})

export const hitboxX = (player: Player): number => player.x

export const hitboxY = (player: Player): number => player.y + 8

export const getPlayerById = (players: HashSet.HashSet<Player>, id: number): Player | null =>Option.getOrNull(Array.findFirst(players, p => p.player_id === id));
// need this for pathfinding
export const getOverlappingCells = (bot: Player): GridCoords[] => {
  const x1 = hitboxX(bot);
  const y1 = hitboxY(bot);
  
  const x2 = x1 + TILE_SIZE - 1;
  const y2 = y1 + TILE_SIZE - 1;

  const corners = [
    [x1, y1], 
    [x2, y1],
    [x1, y2], 
    [x2, y2], 
  ];

  let result: GridCoords[] = [];
  const seen = new Set<string>();

  for (const [px, py] of corners) {
    const cell = pixelToCell(px, py);

    if (cell) {

      const key = `${cell[0]},${cell[1]}`;
      
      if (!seen.has(key)) {
        seen.add(key);
        result = Array.append(result,cell);
      }
    }
  }

  return result;
}

export const isOverlapping = (player: Player, cell: GridCoords): boolean => {
    const TILE_SIZE = 16

    const pCX = player.x + (player.width / 2);
    const pCY = player.y + (player.height / 2);
    const cLeft = cell[1] * TILE_SIZE;
    const cTop = cell[0] * TILE_SIZE;
    
    return (pCX >= cLeft && pCX < cLeft + TILE_SIZE && 
            pCY >= cTop && pCY < cTop + TILE_SIZE);
};
export const getPlayerRow = (player: Player): number => {
  const cx = hitboxX(player) + 16 / 2;
  const cy = hitboxY(player) + 16 / 2;

  const res = pixelToCell(Math.floor(cx), Math.floor(cy));

  if (res) {
    return res[0];
  } else {
    return -1;
  }
};

export const getPlayerCol = (player: Player): number => {
  const cx = hitboxX(player) + 16 / 2;
  const cy = hitboxY(player) + 16 / 2;

  const res = pixelToCell(Math.floor(cx), Math.floor(cy));

  if (res) {
    return res[1];
  } else {
    return -1;
  }
};
